### DISCUSSION TITLE
Invalid model when creating constraints in multiple threads

### DISCUSSION THREAD

**Discussion by sarthakgaur** [google/or-tools#2785] [CP-SAT questions] [answered]
I am using Python and CP-SAT solver. Earlier, I was creating all the constraints in one thread, to improve the performance I thought to create multiple threads and create the constraints in it. Now, when I run the program, it sometimes runs fine, but sometimes I get `MODEL_INVALID` as the status.

All I am doing in these threads is calling the `model.Add()` method. I am also passing some data structures to these threads to read the data from. No write operations are done in these threads. They only read the data from the passed in data structures and create the constraints.

Any ideas why I am getting the `MODEL_INVALID` status sometimes? Any help will be appreciated! Thanks!

**Comment by lperron** [comment, marked as answer]
The model creation API is not thread safe.
